Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,313 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Windows)
+import ucrt
+#elseif canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#elseif canImport(Bionic)
+import Bionic
+#else
+#error("The Byte Buffer module was unable to identify your C library.")
+#endif
+
+/// Describes a way to encode and decode an integer as bytes
+///
+public protocol BinaryIntegerEncodingStrategy {
+    /// Read an integer from a buffer. The reader index should be moved accordingly
+    /// - Returns: The integer
+    func readInteger<IntegerType: FixedWidthInteger>(
+        as: IntegerType.Type,
+        from buffer: inout ByteBuffer
+    ) -> IntegerType?
+
+    /// Write an integer to a buffer. The writer index should be moved accordingly
+    func writeInteger<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        to buffer: inout ByteBuffer
+    ) -> Int
+
+    /// When writing an integer using this strategy, how many bytes we should reserve
+    /// If the actual bytes used by the write function is more or less than this, it will be necessary to shuffle bytes
+    /// Therefore, an accurate prediction here will improve performance
+    var reservedSpaceForInteger: Int { get }
+
+    /// Write an integer to a buffer. The writer index should be moved accordingly
+    /// This function takes  a `reservedSpace` parameter which is the number of bytes we already reserved for writing this integer
+    /// If you use write more or less than this many bytes, we wll need to move bytes around to make that possible
+    /// Therefore, you may decide to encode differently to use more bytes than you actually would need, if it is possible for you to do so
+    /// That way, you waste the reserved bytes, but improve writing performance
+    func writeIntegerWithReservedSpace(
```

Comment:
This API 'stutters' a little bit: could we use `reservedCapacity` to overload `writeInteger`? i.e.

```swift
func writeInteger(_ value: Int, to buffer: inout ByteBuffer, reservedCapacity: Int) -> Int
```